### PR TITLE
Prevent vapor streaming requests from randomly losing a first chunk, stalling the request

### PR DESCRIPTION
**Pull request by Joannis**
Fixes a bug where a streaming body from a request was not processing a chunk, causing the request to stall. This happened when a handler was being set at the exact moment a chunk was being processed, causing the chunk to be added to a now irrelevant array that was just processed while switching to a streaming callback.

See #2906

### CODE REVIEW

**Comment by 0xTim** [comment]
@Joannis still got test failures

**Comment by vzsg** [comment]
Deadlocks are a b—

File: Sources/Vapor/Request/Request+BodyStream.swift [hunk 1]
```diff
@@ -23,58 +23,65 @@ extension Request {
...
         }
 
         func write(_ chunk: BodyStreamResult, promise: EventLoopPromise<Void>?) {
-            // A lock is needed, because some pipeline users (and known unit tests) rely on this being handled synchronously.
             // See https://github.com/vapor/vapor/issues/2906
-            lock.withLockVoid {
+            if self.eventLoop.inEventLoop {
+                write0(chunk, promise: promise)
+            } else {
+                self.eventLoop.execute {
```
Comment (gwynne) [inline Sources/Vapor/Request/Request+BodyStream.swift:38]: It is my understanding that `EventLoop.execute(_:)` already does the "is in event loop" check internally, which in addition, can return false negatives in the normal course of execution - If the problem is that you were executing on the event loop unconditionally before and are skipping the extra "hop", this isn't a guaranteed solution.

Comment (Joannis) [reply Sources/Vapor/Request/Request+BodyStream.swift:38]: It's not supposed to be a guaranteed solution either. The main reason is that I think it's a waste of a hop. And it does _always_ [queue it for the next tick](https://github.com/apple/swift-nio/blob/main/Sources/NIOPosix/SelectableEventLoop.swift#L334). Not only does it fix the test, it also is a good optimisation to have.

Comment (Joannis) [reply Sources/Vapor/Request/Request+BodyStream.swift:38]: Ideally, the test's assumption is to be fixed as well of course.

**Review by 0xTim — APPROVED** [review]
Thanks!

**Pull request merged**